### PR TITLE
Fixes #27512 - Provide way of overriding template snapshots in tests

### PR DESCRIPTION
**Pull request by adamruzicka**
~~For example usage see https://github.com/theforeman/foreman_ansible/pull/296~~

### CODE REVIEW

**Comment by lzap** [comment]
This is a mess, can you provide some description I see strike-through link (is it valid or not?) Is this ready for review?

**Comment by adamruzicka** [comment]
The original link belonged to a previous approach which was abandoned, so it is no longer valid.

The current approach is that for each template there can be multiple variants of a snapshot and the rendered template has to match at least one of the variants to allow the test to pass.

**Comment by ekohl** [comment]
@adamruzicka how about an approach where we set all plugins to disabled in our tests. Then in a plugin you can add tests for snapshots where a certain plugin is enabled.

**Comment by ares** [comment]
@ekohl what is the downside of the current approach in your opinion? I find it simple and in fact like it

**Comment by ekohl** [comment]
Not a strong objection, but I like a bit more control. Silently failing on a snapshot in tests can mean that we ignore one snapshot without knowing.

We currently already have two contexts (safe mode and unsafe mode). I'm wondering if we should introduce another layer of contexts (no plugins, all plugins) by stubbing the `plugin_enabled` macro. It would remove the ambiguity and automatically test a lot of optional code paths increasing coverage.

**Comment by ares** [comment]
I think tests without extensions from plugins would be still ran as part of testing foreman core. Hence broken snapshop would be found. We wouldn't be testing all plugins together, but this allows better granularity imo. Also there may be other impacts of enabling plugin on the template rendering, e.g. when using `respond_to?` instead of `plugin_enabled`.

**Comment by ekohl** [comment]
tl;dr: I've convinced myself that this is the currently best approach, but there's still room for improvement

Could we ship the snapshot as part of foreman_ansible? That foreman core just has a mechanism to specify alternatives but doesn't ship those itself. This could be achieved by looping over the the installed plugins to see if they have a certain directory (`test/fixtures/snapshots` for example).

My main concern is that we'll update templates + built in snapshots but forget the ones with plugins. They get out of date which is then noticed in a plugin. They need to go back to the foreman core repo and get a snapshot updated. By placing the alternative snapshots the impact is reduced.

Now on the other hand a plugin can be valid for multiple Foreman versions but then the snapshots would break. This would also be a problem when a macro is provided by a plugin because then there's nothing really to say about what the content should be.

In the end snapshots are a very brutal way of testing with limited flexibility.

So assuming we can get the tests to pass, I'd be ok with merging this.

**Comment by ares** [comment]
@adamruzicka is the snapshot needed here? It looks you're also adding it in https://github.com/theforeman/foreman_ansible/pull/296

**Comment by adamruzicka** [comment]
So the consensus is the snapshot should live in the plugin?

**Comment by ekohl** [comment]
@adamruzicka either place has its upsides and downsides. I think it living in core has fewer downsides. The sources mostly live in core so the snapshot should too. I still like to have it deterministic so mocking `plugin_present?(plugin)` to always return false is IMHO also an option.

We should also be thinking about variations but storing that all in core creates so much noise. Long term I'd like to turn community-templates into a gem and we could solve it there. There we could also test the rendered output (run `ksvalidator` etc). It would also remove the need to sync files between two repositories.

**Comment by ekohl** [comment]
Looks like this needs a rebase by looking at the failures.

**Comment by adamruzicka** [comment]
Rebased

**Review by lzap — APPROVED** [review]
Works fine, render test is passing locally.

**Comment by lzap** [comment]
I will let @ekohl to merge since he did most of the review.

**Comment by ekohl** [comment]
@adamruzicka this is currently two commits. Should those be squashed?

**Comment by adamruzicka** [comment]
Yes

**Comment by ekohl** [comment]
Thanks!
